# MATH 221: Assignment 5
<h3><font color= red>  Due Apr. 14  </font></h3>
<h4> Up to three students may submit the same notebook.  Enter the names of all students who contributed to the assignment in the next cell.</h4>
<h4> Please also include your names in the name of the notebook file you submit to Blackboard.</h4>
<br><br>

## Names:

<br>

In [2]:
## IMPORTANT!!  Execute this import cell each time before you begin working.
## Before you begin, transfer the laguide.py file from Blackboard to the JupyterHub.

import numpy as np
import laguide as lag


<br><br><br>

### Power method estimates for eigenvalues


In the first problem, you will use the Power Method to esimate the eigenvector associated with largest eigenvalue of the matrix $A$.

$$
\begin{equation}
A = \left[ \begin{array}{rrrrr}
2 & -1 & -1 & 1 \\  
-1 & 4 & -3 & 2 \\
-1 & -3 & 0 & 2 \\
1 & 2 & 2  & 6
\end{array} \right] 
\end{equation}
$$



In [3]:
A = np.array([[2,-1,-1,1],[-1,4,-3,2],[-1,-3,0,2],[1,2,2,6]])


The idea is to build a sequence of vectors $\{x_0$, $x_1$, $x_2$, ...\} by multiplying by $A$ and scaling to unit length.  We'll choose a vector at random to start the sequence.

In [4]:
## Choose something here for the intial x_0

x = np.array([[1],[1],[1],[1]])
print(x)

[[1]
 [1]
 [1]
 [1]]


Starting with $x_0$, we multiply by $A$ and scale the result to unit length to find $x_1$.

In [5]:
x = A@x
x = x/(lag.Magnitude(x))
print(x)

[[ 0.0877058 ]
 [ 0.1754116 ]
 [-0.1754116 ]
 [ 0.96476382]]


Repeat to find $x_2$ from $x_1$.

In [6]:
x = A@x
x = x/(lag.Magnitude(x))
print(x)

[[0.1663389 ]
 [0.44783551]
 [0.19192951]
 [0.85728513]]


Instead of reapeating the code, we tuck these two commands inside of a loop that runs $k$ times.  We could print out each $x_k$ but that might be too much output if $k$ is large.  Instead, maybe we just print the final result.


In [7]:
## Choose a starting x_0
x = np.array([[1],[1],[1],[1]])
## Choose the number of steps to take in the sequence
k = 4
for i in range(k):
    x = A@x
    x = x/lag.Magnitude(x)


### Problem 1a:

Experiment to determine how large $k$ must be so that $x_k$ and $x_{k+1}$ are approximately the same.  Specifically, find a $k$ so that $||x_{k+1}-x_k||< 10^{-8}$.  Demonstrate with a computation.



<h3><font color=blue> Answer 1a </font> </h3>



In [8]:
## Choose a starting x_0
x = np.array([[1],[1],[1],[1]])
## Choose the number of steps to take in the sequence
k = 51

for i in range(k):
    x = A@x
    x = x/lag.Magnitude(x)

    
## Calculate one addional vector x_{k+1} and compare it with x_{k}    
x_new = A@x
x_new = x_new/lag.Magnitude(x_new)
print(lag.Magnitude(x-x_new))

9.727784506066492e-09


$k$ needs to be at larger than 50 in this case.

<br><br>

### Problem 1b:

The vector from the previous computation must be an approximate eigenvector of $A$.  Use the eigenvector to estimate the associated eigenvalue of $A$.  


<br><br><br>

### Answer 1b:

If $x$ is an eigenvector (so $Ax=\lambda x$) and $||x||=1$, then 

$$
\begin{equation}
||Ax|| = ||\lambda x|| = \lambda||x|| = \lambda
\end{equation}
$$

In [9]:
## CODE HERE
print(lag.Magnitude(A@x))

7.261802245259969


<br><br><br>

### Problem 1c

Now if $A$ is invertible and $\lambda$ is an eigenvalue so that $Ax=\lambda x$, then $\frac{1}{\lambda}$ must be an eigenvalue of $A^{-1}$ since $x = \lambda A^{-1}x$.  This means that the smallest eigenvalue of $A$ is the *largest* eigenvalue of $A^{-1}$.  Use this idea to estimate the smallest eigenvalue of $A$ using the Power Method.  **Verify your answer** by checking that $Ax=\lambda x$ for your $\lambda$ and $x$.

<h3><font color=blue> Answer 1c </font> </h3>



In [10]:
## CODE HERE

## Calculate A inverse
A_inv = lag.Inverse(A)
## Choose a starting x_0
x = np.array([[1],[1],[1],[1]])
## Choose the number of steps to take in the sequence
k = 100

for i in range(k):
    x = A_inv@x
    x = x/lag.Magnitude(x)

## Calculate one addional vector x_{k+1} and compare it with x_{k}    
x_new = A@x
x_new = x_new/lag.Magnitude(x_new)
print(lag.Magnitude(x-x_new),'\n')

print("Eigenvalue of A inverse is approximately: ",lag.Magnitude(A_inv@x))
print("Eigenvalue of A is approximately: ",1/lag.Magnitude(A_inv@x))

1.0704140406482906e-09 

Eigenvalue of A inverse is approximately:  0.4169583867171188
Eigenvalue of A is approximately:  2.398320868116846


Verify that this vector is an eigenvector of $A$.

In [11]:
eig_value_A = 1/lag.Magnitude(A_inv@x)

## Compare Ax with lambda*x
print(A@x)
print(eig_value_A*x)

[[ 2.23490926]
 [-0.1123194 ]
 [-0.85902571]
 [-0.08113411]]
[[ 2.23490927]
 [-0.1123194 ]
 [-0.85902571]
 [-0.08113411]]


In [12]:
## Check our results using SciPy function for computing eigenvalues

import scipy.linalg as sla

evals,evects = sla.eig(A)
print(evals)

[-3.        +0.j  2.39832087+0.j  5.33987689+0.j  7.26180225+0.j]


The smallest eigenvalue of $A$ is approximately 2.3983, and the largest eigenvalue is approximately 7.2618.

<br><br><br>

### Problem 2a

Suppose that a survey estimates that each year in BC 15\% of Liberals switch to the Green Party, and 20\% switch
to the NDP.  The surve also estimates that 20\% of NDP members switch to the Liberals, 15\% of NDP members switch to the Green Party,
10\% of Green Party members change to the NDP, and 10\% of Green Party members switch to the Liberals.

Let $N_k$, $G_k$, and $L_k$ be the proportions of NDP, Liberal, and Green members at year $k$, and let $x$ be the vector with
these components.  Write down the transition matrix $A$ so that $Ax_k=x_{k-1}$.

<h3><font color=blue> Answer 2a </font> </h3>


$$
\begin{equation}
A = \left[ \begin{array}{rrr}  
0.65 & 0.1 & 0.2  \\
0.15 & 0.8 & 0.15  \\
0.2 & 0.1 & 0.65   
\end{array} \right] 
\end{equation}
$$


<br><br><br>

### Problem 2b

Find the equilibrium proportion of $N_k$, $G_k$, and $L_k$ and verify that it is an equilbrium with a computation.

<br>
<br>

<h3><font color=blue> Answer 2b </font> </h3>


Solve $(A-I)v=0$

In [13]:
A = np.array([[0.65,0.1,0.2],[0.15,0.8,0.15],[0.2,0.1,0.65]])
I = np.eye(3)
print(lag.FullRowReduction(A-I))

[[ 1.   0.  -1. ]
 [ 0.   1.  -1.5]
 [ 0.   0.   0. ]]


The free variable means we have a choice.  One possible solution is:


$$
\begin{equation}
v = \left[ \begin{array}{r}  
2  \\
3  \\
2   
\end{array} \right] 
\end{equation}
$$

Since we are looking for propotions of the population, we scale down the total of the entries.

$$
\begin{equation}
v = \left[ \begin{array}{r}  
\frac27  \\
\frac37  \\
\frac27   
\end{array} \right] 
\end{equation}
$$

Now we check that $Av = v$.

In [14]:
v = np.array([[2/7],[3/7],[2/7]])
print(A@v)
print(v)

[[0.28571429]
 [0.42857143]
 [0.28571429]]
[[0.28571429]
 [0.42857143]
 [0.28571429]]
